File: frontend/src/types/asset.ts
// 资产类型定义，映射后端 models/movies/asset_models.py 与 routers/schemas/asset.py

export enum AssetType {
  VIDEO = 'video',
  SUBTITLE = 'subtitle',
  IMAGE = 'image',
}

export enum AssetStoreType {
  LOCAL = 'local',
  S3 = 's3',
}

export interface VideoMetadata {
  type: AssetType.VIDEO;
  size: number;
  duration?: number;
  quality?: string;
  codec?: string;
  has_thumbnail?: boolean;
  has_sprite?: boolean;
  width?: number;
  height?: number;
}

export interface SubtitleMetadata {
  type: AssetType.SUBTITLE;
  size: number;
  language?: string;
}

export interface ImageMetadata {
  type: AssetType.IMAGE;
  size: number;
  width?: number;
  height?: number;
}

export type AssetMetadata = VideoMetadata | SubtitleMetadata | ImageMetadata;

export interface AssetBase {
  library_id: string;
  movie_id: string;
  type: AssetType;
  name: string; // 播放时显示
  path: string;
  store_type: AssetStoreType;
  actual_path?: string;
  description: string;
  tags?: string[];
}

export type AssetCreate = AssetBase

export interface AssetUpdate {
  name?: string;
  description?: string;
  actual_path?: string;
  tags?: string[];
  metadata?: AssetMetadata;
}

export interface AssetInDB extends AssetBase {
  id: string;
  metadata?: AssetMetadata;
  is_deleted: boolean;
  deleted_at?: string | null;
  created_at: string;
  updated_at: string;
}

export type AssetRead = AssetInDB

export interface AssetPageResult {
  items: AssetRead[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

// 路由层 Schema（routers/schemas/asset.py）
export type AssetCreateRequestSchema = Omit<AssetCreate, "path" | "library_id">;

export type AssetUpdateRequestSchema = AssetUpdate

export type AssetReadResponseSchema = AssetRead

export type AssetPageResultResponseSchema = AssetPageResult

// 与后端 routers/schemas/asset.py 新增保持一致
export interface AssetBatchUpdateRequestSchema {
  asset_ids: string[];
  patch: AssetUpdateRequestSchema;
}

File: backend/src/routers/schemas/asset.py
from typing import List, Optional, Union
from pydantic import BaseModel, Field
from src.models import AssetCreate, AssetUpdate, AssetRead, AssetPageResult, AssetStoreType


class AssetCreateRequestSchema(AssetCreate):
    """创建资产请求模型"""

    # 排除 path 和 library_id 字段
    path: Optional[str] = Field(None, description="文件路径", exclude=True)
    library_id: Optional[str] = Field(None, description="库ID", exclude=True)
    store_type: Optional[AssetStoreType] = Field(AssetStoreType.LOCAL, description="存储类型")

class AssetUpdateRequestSchema(AssetUpdate):
    """更新资产请求模型"""

    pass  # 所有字段从 AssetUpdate 继承


class AssetReadResponseSchema(AssetRead):
    """读取资产响应模型"""

    pass  # 所有字段从 AssetRead 继承，Pydantic 模型之间的兼容性机制 + FastAPI 的自动响应序列化机制，可直接 return AssetRead


class AssetPageResultResponseSchema(AssetPageResult):
    """资产分页结果响应模型"""

    pass

# 新增：批量更新请求模型
class AssetBatchUpdateRequestSchema(BaseModel):
    asset_ids: List[str] = Field(..., description="资产ID列表")
    patch: AssetUpdateRequestSchema = Field(..., description="更新内容")
